### ISSUE TITLE
speeding up deep graphs

### ISSUE THREAD

**Issue by ekg** [marschall-lab/GFAffix#2]
On deep graphs (2k-fold) I'm seeing gfaffix taking quite a bit of time. It's essentially single threaded, right? Is there a possible way to adapt it to operate in parallel?

**Comment by danydoerr** [comment]
There is potential for speedup by multithreading parts of GFAffix, especially if the bottle neck is the graph traversal. I'd be interested in knowing which step in the algorithm affects the running time on these deep graphs. The graph traversal can be parallelized, but it's  not embarrassingly parallelizable, as the graph editing (which is intertwined in the graph traversal) can only be done safely in a single thread.

**Comment by natir** [comment]
If you give me some sample dataset I can try to perform runtime analysis and try to improve running time by parallelized or not.

**Comment by danydoerr** [comment]
@ekg do you have such a "deep graph" handy?

**Comment by natir** [comment]
Or just smaller but similar graph.

**Comment by ekg** [comment]
Yes, but they are big. I don't think you need to use a deep graph to see if
parallelism can help. Probably a chunk of the MHC will be enough to
profile, or any graph that you've already used for testing that takes e.g.
5-10s to process.

**Comment by danydoerr** [comment]
Do these deep graphs have a particular high node degree, or is the number of nodes exceptionally high?

**Comment by ekg** [comment]
High node degree and path depth. 2k fold coverage

**Comment by danydoerr** [comment]
GFAffix should scale linearly with path depth, overall this shouldn't dominate the runtime. But node degree is certainly a bottle neck.

**Comment by natir** [comment]
@ekg Could you give me advise how to build a similar graph, type of data, pipeline, tools, etc…

**Comment by ekg** [comment]
@natir sorry to take a while here. I may need to share this somehow, but the graph is rather large and will take 4 days to build if you do so from scratch. I'll see if I can get a simpler test case together.

**Comment by danydoerr** [comment]
@ekg Just so you know that your issue is not forgotten: I have a solution for parallelizing GFAffix and will work on this sometime soonish.

**Comment by AndreaGuarracino** [comment]
@danydoerr any updates on the parallelization?

**Comment by danydoerr** [comment]
@AndreaGuarracino thanks for asking. I'm preparing a release that should be out in the next few days. Do you want to have already a binary to test on? I'm also changing the cli...

**Comment by AndreaGuarracino** [comment]
Of course, please 'exit' the binary xD

________________________________

**Comment by danydoerr** [comment]
@AndreaGuarracino  voilà [gfaffix-0.2.0-prerelease.tar.gz](https://github.com/user-attachments/files/18213167/gfaffix-0.2.0-prerelease.tar.gz)

**Comment by AndreaGuarracino** [comment]
Thx! Now I've noticed your `refined_deduplication` branch!

Looks extremely great. With 1 thread it is already a bit faster than the current `main` branch. And it is also slimmer in memory.

**Comment by danydoerr** [comment]
Yes. That is the current development branch from which I generated the binary. There should be a small speed up when using up to 4 threads (`-p4`)

**Comment by danydoerr** [comment]
@AndreaGuarracino do you have ungfaffixed graphs of the new HPRC assemblies that I can test on?

**Comment by AndreaGuarracino** [comment]
Not yet, they'll come soon. I've started using the new gfaffix everywhere, so it is already "in production".

________________________________

**Comment by danydoerr** [comment]
New version is released now. It still requires improvement in speed and memory usage for these large graphs. The bottleneck is now only the i/o part of gfaffix, and I already have some concrete ideas how to improve it. I'll leave this issue open for now.
